src/ccl_halomod.c

```
@@ -0,0 +1,310 @@
+// TODO: Actually code this up
+static double z_formation_Bullock(ccl_cosmology *cosmo, double halomass, double a, int *status){ 
+  return 0.;
+}
+
+// The non-linear mass (sigma(M*)=delta_c))
+// TODO: Actually code this up
+static double Mstar(){
+  return 1e13;
+}
+
+// The concentration-mass relation for haloes
+// TODO: make consistency check so that ccl_halo_concentration only runs if called with appropriate definition
+// TODO: e.g. if Delta != 200 rho_{mean}, should not function (or should it?)
+// TODO: should this be moved to the ccl_massfunc.c ?
+double ccl_halo_concentration(ccl_cosmology *cosmo, double halomass, double a, int * status)
+{
+
+  // Set concentration-mass relation
+  // 1 - Bhattaharya et al. (2011)
+  // 2 - Full Bullock et al. (2001)
+  // 3 - Virial Duffy et al. (2008)
+  // 4 - Constant concentration (useful for testing)
+  // 5 - Simple Bullock et al. (2001)
+  int iconc=3;
+  
+  // Bhattacharya et al. 2011, Delta = 200 rho_{mean} (Table 2)
+  if(iconc==1){    
+    double gz = ccl_growth_factor(cosmo,a,status);
+    double g0 = ccl_growth_factor(cosmo,1.0,status);
+    return 9.*pow(ccl_nu(cosmo,halomass,a,status),-0.29)*pow(gz/g0,1.15);
+  }
+
+  // Full Bullock et al. (2001)
+  else if(iconc==2){    
+    double A = 4.;
+    double z = -1.+1./a;
+    double zf = z_formation_Bullock(cosmo,halomass,a,status);
+    return A*(1.+zf)/(1.+z);
+  }
+
+  // Duffy et al. (2008; 0804.2486; Table 1, second section: Delta = Virial)
+  else if(iconc==3){
+    double Mpiv = 2e12/cosmo->params.h; // Pivot mass in Msun (note in the paper units are Msun/h)
+    double A = 7.85;
+    double B = -0.081;
+    double C = -0.71;
+    return A*pow(halomass/Mpiv,B)*pow(a,-C); 
+  }
+
+  // Constant concentration (good for tests)
+  else if(iconc==4){
+    return 4.;
+  }
+
+  // Simple Bullock et al. (2001) relation
+  else if(iconc==5){
+    return 9.*pow(halomass/Mstar(),-0.13);
+  }
+
+  // Something went wrong
+  else{    
+    exit(0);
```

This should raise an error. See `ccl_error.c` for helper functions to help you do that. Here are the essentials:

1. Define a new error variable in `ccl_error.h`, e.g. `#define CCL_ERROR_HALOCONC xxx`
1. Set the status variable `*status = CCL_ERROR_HALOCONC;`
1. Raise an error, using `ccl_raise_exception(int err, char* msg)`.
1. Return `NAN`.

Calling `exit()` is bad because it crashes the Python interpreter :)

---

Done